libbpfgo.go

```diff
@@ -1331,6 +1384,61 @@ func (p *BPFProg) AttachCgroup(cgroupV2DirPath string) (*BPFLink, error) {
 	return bpfLink, nil
 }
 
+// AttachCgroupLegacy attaches the BPFProg to a cgroup described by given fd.
+// It first tries to use the most recent attachment method and, if that does
+// not work, instead of failing it tries the legacy way: to attach the cgroup
+// eBPF program without previously creating a link. Related to upstream kernel
+// commit af6eea57437a ("bpf: Implement bpf_link-based cgroup BPF program
+// attachment"). This allows attaching cgroup eBPF ingress/egress in older
+// kernels.
+func (p *BPFProg) AttachCgroupLegacy(cgroupV2DirPath string, attachType BPFAttachType) (*BPFLink, error) {
+	// bpfLink, err := p.AttachCgroup(cgroupV2DirPath)
+	// if err == nil {
+	// 	return bpfLink, nil
+	// }
+	// Try the legacy attachment method before fully failing
+	cgroupDirFD, err := getCgroupDirFD(cgroupV2DirPath)
+	if err != nil {
+		return nil, err
+	}
+	defer syscall.Close(cgroupDirFD)
+	progFD := C.bpf_program__fd(p.prog)
+	ret := C.bpf_prog_attach_cgroup_legacy(progFD, C.int(cgroupDirFD), C.int(attachType))
+	if ret < 0 {
+		return nil, fmt.Errorf("failed to attach (legacy) program %s to cgroupv2 %s", p.name, cgroupV2DirPath)
+	}
+	dirName := strings.ReplaceAll(cgroupV2DirPath[1:], "/", "-")
+	fakeBpfLink := &BPFLink{
+		link:      nil, // detach/destroy made with progfd
+		prog:      p,
+		eventName: fmt.Sprintf("cgroup-%s-%s", p.name, dirName),
+		// info bellow needed for detach (there isn't a real ebpf link)
+		linkType:  CgroupLegacy, 
+		attachType: attachType,
+		cgroupDir: cgroupV2DirPath,
+	}
+	return fakeBpfLink, nil
+}
+
+// DetachCgroupLegacy detaches the BPFProg from a cgroup described by given fd.
+// This is needed because in legacy attachment there is no BPFLink, just a fake
+// one (kernel did not support it, nor libbpf). This function should be called
+// by the (*BPFLink)->Destroy() function, since BPFLink is emulated (so user
+// don´t need to distinguish between regular and legacy cgroup detachments).
```

Comment:
```suggestion
// DetachCgroupLegacy detaches the BPFProg from a cgroup described by the given fd.
// This is needed because in legacy attachment there is no BPFLink, just a fake
// one (kernel did not support it, nor libbpf). This function should be called
// by the (*BPFLink)->Destroy() function, since BPFLink is emulated (so users
// don´t need to distinguish between regular and legacy cgroup detachments).
```